src/hazelcore/extractions/ocaml/OCamlExtraction_Pat.re

```diff
@@ -0,0 +1,114 @@
+// mark with a constructor to tell apart with typ, exp
+type t =
+  | OCamlPat(string);
+
+exception Pat_Hole;
+exception Pat_Keyword(ExpandingKeyword.t); //there's Keyword
+exception Pat_Invalid(string);
+
+// the ocaml naming standard can be found in Chapter 7.3 of ocaml-manual (https://caml.inria.fr/pub/docs/manual-ocaml/names.html)
+let ocaml_name_check = (name: Var.t): bool => {
+  let re = Re.Str.regexp("^[_a-z][_a-zA-Z0-9]*$");
+  Re.Str.string_match(re, name, 0);
+};
+
+let rec extract = (dp: DHPat.t): t =>
```

Comment:
Can `update_pattern` and `extract` be combined into one function?

Reply:
It can be combined with `extract`, but may cause the `extract` function to return a more complicated result to destruct. However, the `update_pattern` is only used for a limited few times, so I think it's better to separate them.

Reply:
Except for internal recursions, `OCamlExtraction_pat.extract` and `update_pattern` are always used together.  Also, the internal recursions will line up once they are the same function.

There are also philosophical reasons that these are the same.  You always have to update the scope when extracting from a pattern.

Reply:
Oh, it seems a good reason. So may I just append it with `and`, or combine the ctx into the return result of `OCamlExtraction_exp.extract`?